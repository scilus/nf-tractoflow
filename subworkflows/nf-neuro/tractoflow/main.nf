
// PREPROCESSING
include {   PREPROC_DWI                                               } from '../preproc_dwi/main'
include {   PREPROC_T1                                                } from '../preproc_t1/main'
include {   RECONST_DTIMETRICS as REGISTRATION_FA                     } from '../../../modules/nf-neuro/reconst/dtimetrics/main'
include {   REGISTRATION as T1_REGISTRATION                           } from '../registration/main'
include {   REGISTRATION_ANTSAPPLYTRANSFORMS as TRANSFORM_WMPARC      } from '../../../modules/nf-neuro/registration/antsapplytransforms/main'
include {   REGISTRATION_ANTSAPPLYTRANSFORMS as TRANSFORM_APARC_ASEG  } from '../../../modules/nf-neuro/registration/antsapplytransforms/main'
include {   REGISTRATION_ANTSAPPLYTRANSFORMS as TRANSFORM_LESION_MASK } from '../../../modules/nf-neuro/registration/antsapplytransforms/main'
include {   REGISTRATION_ANTSAPPLYTRANSFORMS as TRANSFORM_BRAINMASK   } from '../../../modules/nf-neuro/registration/antsapplytransforms/main'
include {   ANATOMICAL_SEGMENTATION                                   } from '../anatomical_segmentation/main'

// RECONSTRUCTION
include {   RECONST_FRF        } from '../../../modules/nf-neuro/reconst/frf/main'
include {   RECONST_MEANFRF    } from '../../../modules/nf-neuro/reconst/meanfrf/main'
include {   RECONST_DTIMETRICS } from '../../../modules/nf-neuro/reconst/dtimetrics/main'
include {   RECONST_FODF       } from '../../../modules/nf-neuro/reconst/fodf/main'

// TRACKING
include { TRACKING_PFTTRACKING   } from '../../../modules/nf-neuro/tracking/pfttracking/main'
include { TRACKING_LOCALTRACKING } from '../../../modules/nf-neuro/tracking/localtracking/main'


// ** UTILITY FUNCTIONS ** //

def group_frf ( label, ch_frf ) {
    return ch_frf
        .map{ _meta, frf -> frf }
        .flatten()
        .map{ frf_list -> [label, frf_list] }
}


workflow TRACTOFLOW {
    take:
        ch_dwi              // channel : [required] meta, dwi, bval, bvec
        ch_t1               // channel : [required] meta, t1
        ch_sbref            // channel : [optional] meta, sbref
        ch_rev_dwi          // channel : [optional] meta, rev_dwi, rev_bval, rev_bvec
        ch_rev_sbref        // channel : [optional] meta, rev_sbref
        ch_wmparc           // channel : [optional] meta, wmparc
        ch_aparc_aseg       // channel : [optional] meta, aparc_aseg
        ch_topup_config     // channel : [optional] topup_config
        ch_bet_template     // channel : [optional] meta, bet_template
        ch_bet_probability  // channel : [optional] meta, bet_probability
        ch_lesion_mask      // channel : [optional] meta, lesion_mask
    main:

        ch_versions = Channel.empty()

        /* PREPROCESSING */

        //
        // SUBWORKFLOW: Run PREPROC_DWI
        //
        PREPROC_DWI(
            ch_dwi,
            ch_rev_dwi,
            ch_sbref,
            ch_rev_sbref,
            ch_topup_config
        )
        ch_versions = ch_versions.mix(PREPROC_DWI.out.versions.first())

        //
        // SUBWORKFLOW: Run PREPROC_T1
        //
        PREPROC_T1(
            ch_t1,
            ch_bet_template,
            ch_bet_probability,
            Channel.empty(),
            Channel.empty(),
            Channel.empty(),
            Channel.empty()
        )
        ch_versions = ch_versions.mix(PREPROC_T1.out.versions.first())

        /* RECONSTRUCTION - PART I - doesn't need anatomy */

        //
        // MODULE: Run RECONST/DTIMETRICS
        //
        ch_dti_metrics = PREPROC_DWI.out.dwi
            .join(PREPROC_DWI.out.bval)
            .join(PREPROC_DWI.out.bvec)
            .join(PREPROC_DWI.out.b0_mask)

        RECONST_DTIMETRICS( ch_dti_metrics )
        ch_versions = ch_versions.mix(RECONST_DTIMETRICS.out.versions.first())


        //
        // SUBWORKFLOW: Run REGISTRATION
        //
        T1_REGISTRATION(
            PREPROC_T1.out.t1_final,
            PREPROC_DWI.out.b0,
            RECONST_DTIMETRICS.out.fa,
            Channel.empty(),
            Channel.empty(),
            Channel.empty()
        )
        ch_versions = ch_versions.mix(T1_REGISTRATION.out.versions.first())

        /* SEGMENTATION */

        //
        // MODULE: Run REGISTRATION_ANTSAPPLYTRANSFORMS (TRANSFORM_BRAINMASK)
        //
        TRANSFORM_BRAINMASK(
            PREPROC_T1.out.mask_final
                .join(PREPROC_DWI.out.b0)
                .join(T1_REGISTRATION.out.transfo_image)
        )
        ch_versions = ch_versions.mix(TRANSFORM_BRAINMASK.out.versions.first())

        //
        // MODULE: Run REGISTRATION_ANTSAPPLYTRANSFORMS (TRANSFORM_WMPARC)
        //
        TRANSFORM_WMPARC(
            ch_wmparc
                .join(PREPROC_DWI.out.b0)
                .join(T1_REGISTRATION.out.transfo_image)
        )
        ch_versions = ch_versions.mix(TRANSFORM_WMPARC.out.versions.first())

        //
        // MODULE: Run REGISTRATION_ANTSAPPLYTRANSFORMS (TRANSFORM_APARC_ASEG)
        //
        TRANSFORM_APARC_ASEG(
            ch_aparc_aseg
                .join(PREPROC_DWI.out.b0)
                .join(T1_REGISTRATION.out.transfo_image)
        )
        ch_versions = ch_versions.mix(TRANSFORM_APARC_ASEG.out.versions.first())

        //
        // Module: Run REGISTRATION_ANTSAPPLYTRANSFORMS (TRANSFORM_LESION_MASK)
        TRANSFORM_LESION_MASK(
            ch_lesion_mask
                .join(PREPROC_DWI.out.b0)
                .join(T1_REGISTRATION.out.transfo_image)
        )
        ch_versions = ch_versions.mix(TRANSFORM_LESION_MASK.out.versions.first())

        //
        // SUBWORKFLOW: Run ANATOMICAL_SEGMENTATION
        //
        ANATOMICAL_SEGMENTATION(
            T1_REGISTRATION.out.image_warped,
            TRANSFORM_WMPARC.out.warped_image
                .join(TRANSFORM_APARC_ASEG.out.warped_image),
            TRANSFORM_LESION_MASK.out.warped_image,
            Channel.empty()
        )
        ch_versions = ch_versions.mix(ANATOMICAL_SEGMENTATION.out.versions.first())

        /* RECONSTRUCTION - PART II - needs anatomy */

        //
        // MODULE: Run RECONST/FRF
        //
        ch_reconst_frf = PREPROC_DWI.out.dwi
            .join(PREPROC_DWI.out.bval)
            .join(PREPROC_DWI.out.bvec)
            .join(PREPROC_DWI.out.b0_mask)
            .join(ANATOMICAL_SEGMENTATION.out.wm_mask)
            .join(ANATOMICAL_SEGMENTATION.out.gm_mask)
            .join(ANATOMICAL_SEGMENTATION.out.csf_mask)

        RECONST_FRF( ch_reconst_frf )
        ch_versions = ch_versions.mix(RECONST_FRF.out.versions.first())

        /* Run fiber response averaging over subjects */
        ch_single_frf = RECONST_FRF.out.frf
            .map{ it + [[], []] }

        ch_fiber_response = RECONST_FRF.out.wm_frf
            .join(RECONST_FRF.out.gm_frf)
            .join(RECONST_FRF.out.csf_frf)
            .mix(ch_single_frf)

        if ( params.frf_average_from_data ) {
            ch_single_frf = group_frf("ssst", RECONST_FRF.out.frf)

            ch_wm_frf = group_frf("wm", RECONST_FRF.out.wm_frf)
            ch_gm_frf = group_frf("gm", RECONST_FRF.out.gm_frf)
            ch_csf_frf = group_frf("csf", RECONST_FRF.out.csf_frf)

            ch_meanfrf = ch_single_frf
                .mix(ch_wm_frf)
                .mix(ch_gm_frf)
                .mix(ch_csf_frf)

            RECONST_MEANFRF( ch_meanfrf )
            ch_versions = ch_versions.mix(RECONST_MEANFRF.out.versions.first())

            ch_meanfrf = RECONST_MEANFRF.out.meanfrf
                .map{ ["frf"] + it }
                .branch{
                    ssst: it[1] == "ssst"
                    wm: it[1] == "wm"
                    gm: it[1] == "gm"
                    csf: it[1] == "csf"
                }

            ch_fiber_response = ch_meanfrf.wm.map{ [it[0], it[2]] }
                .join(ch_meanfrf.gm.map{ [it[0], it[2]] })
                .join(ch_meanfrf.csf.map{ [it[0], it[2]] })
                .map{ it[1..-1] }
                .mix(ch_meanfrf.ssst.map{ [it[1], [], []] })
                .combine(RECONST_FRF.out.map{ it[0] })
        }

        //
        // MODULE: Run RECONST/FODF
        //
        ch_reconst_fodf = PREPROC_DWI.out.dwi
            .join(PREPROC_DWI.out.bval)
            .join(PREPROC_DWI.out.bvec)
            .join(PREPROC_DWI.out.b0_mask)
            .join(RECONST_DTIMETRICS.out.fa)
            .join(RECONST_DTIMETRICS.out.md)
            .join(ch_fiber_response)
        RECONST_FODF( ch_reconst_fodf )
        ch_versions = ch_versions.mix(RECONST_FODF.out.versions.first())

        //
        // MODULE: Run TRACKING/PFTTRACKING
        //
        ch_pft_tracking = Channel.empty()
        if ( params.run_pft ) {
            ch_input_pft_tracking = ANATOMICAL_SEGMENTATION.out.wm_mask
                .join(ANATOMICAL_SEGMENTATION.out.gm_mask)
                .join(ANATOMICAL_SEGMENTATION.out.csf_mask)
                .join(RECONST_FODF.out.fodf)
                .join(RECONST_DTIMETRICS.out.fa)
            TRACKING_PFTTRACKING( ch_input_pft_tracking )
            ch_versions = ch_versions.mix(TRACKING_PFTTRACKING.out.versions.first())

            ch_pft_tracking = TRACKING_PFTTRACKING.out.trk
                .join(TRACKING_PFTTRACKING.out.config)
                .join(TRACKING_PFTTRACKING.out.includes)
                .join(TRACKING_PFTTRACKING.out.excludes)
                .join(TRACKING_PFTTRACKING.out.seeding)
        }

        //
        // MODULE: Run TRACKING/LOCALTRACKING
        //
        ch_local_tracking = Channel.empty()
        if ( params.run_local_tracking ) {
            ch_input_local_tracking = ANATOMICAL_SEGMENTATION.out.wm_mask
                .join(RECONST_FODF.out.fodf)
                .join(RECONST_DTIMETRICS.out.fa)
            TRACKING_LOCALTRACKING( ch_input_local_tracking )
            ch_versions = ch_versions.mix(TRACKING_LOCALTRACKING.out.versions.first())

            ch_local_tracking = TRACKING_LOCALTRACKING.out.trk
                .join(TRACKING_LOCALTRACKING.out.config)
                .join(TRACKING_LOCALTRACKING.out.seedmask)
                .join(TRACKING_LOCALTRACKING.out.trackmask)
        }

    emit:

        // IN DIFFUSION SPACE
        dwi                     = PREPROC_DWI.out.dwi
                                    .join(PREPROC_DWI.out.bval)
                                    .join(PREPROC_DWI.out.bvec)
        t1                      = T1_REGISTRATION.out.image_warped
        brain_mask              = TRANSFORM_BRAINMASK.out.warped_image
        wm_mask                 = ANATOMICAL_SEGMENTATION.out.wm_mask
        gm_mask                 = ANATOMICAL_SEGMENTATION.out.gm_mask
        csf_mask                = ANATOMICAL_SEGMENTATION.out.csf_mask
        wm_map                  = ANATOMICAL_SEGMENTATION.out.wm_map
        gm_map                  = ANATOMICAL_SEGMENTATION.out.gm_map
        csf_map                 = ANATOMICAL_SEGMENTATION.out.csf_map
        aparc_aseg              = TRANSFORM_APARC_ASEG.out.warped_image
        wmparc                  = TRANSFORM_WMPARC.out.warped_image

        // REGISTRATION
        anatomical_to_diffusion = T1_REGISTRATION.out.transfo_image
        diffusion_to_anatomical = T1_REGISTRATION.out.transfo_trk

        // IN ANATOMICAL SPACE
        t1_native               = PREPROC_T1.out.t1_final

        // DTI
        dti_tensor              = RECONST_DTIMETRICS.out.tensor
        dti_md                  = RECONST_DTIMETRICS.out.md
        dti_rd                  = RECONST_DTIMETRICS.out.rd
        dti_ad                  = RECONST_DTIMETRICS.out.ad
        dti_fa                  = RECONST_DTIMETRICS.out.fa
        dti_rgb                 = RECONST_DTIMETRICS.out.rgb
        dti_peaks               = RECONST_DTIMETRICS.out.evecs_v1
        dti_evecs               = RECONST_DTIMETRICS.out.evecs
        dti_evals               = RECONST_DTIMETRICS.out.evals
        dti_residual            = RECONST_DTIMETRICS.out.residual
        dti_ga                  = RECONST_DTIMETRICS.out.ga
        dti_mode                = RECONST_DTIMETRICS.out.mode
        dti_norm                = RECONST_DTIMETRICS.out.norm

        // FODF
        fiber_response          = ch_fiber_response
        fodf                    = RECONST_FODF.out.fodf
        wm_fodf                 = RECONST_FODF.out.wm_fodf
        gm_fodf                 = RECONST_FODF.out.gm_fodf
        csf_fodf                = RECONST_FODF.out.csf_fodf
        fodf_rgb                = RECONST_FODF.out.vf_rgb
        fodf_peaks              = RECONST_FODF.out.peaks
        afd_max                 = RECONST_FODF.out.afd_max
        afd_total               = RECONST_FODF.out.afd_total
        afd_sum                 = RECONST_FODF.out.afd_sum
        nufo                    = RECONST_FODF.out.nufo
        volume_fraction         = RECONST_FODF.out.vf

        // TRACKING
        pft_tractogram          = ch_pft_tracking.map{ [it[0], it[1]] }
        pft_config              = ch_pft_tracking.map{ [it[0], it[2]] }
        pft_map_include         = ch_pft_tracking.map{ [it[0], it[3]] }
        pft_map_exclude         = ch_pft_tracking.map{ [it[0], it[4]] }
        pft_seeding_mask        = ch_pft_tracking.map{ [it[0], it[5]] }
        local_tractogram        = ch_local_tracking.map{ [it[0], it[1]] }
        local_config            = ch_local_tracking.map{ [it[0], it[2]] }
        local_seeding_mask      = ch_local_tracking.map{ [it[0], it[3]] }
        local_tracking_mask     = ch_local_tracking.map{ [it[0], it[4]] }

        // QC
        nonphysical_voxels      = RECONST_DTIMETRICS.out.nonphysical
        pulsation_in_dwi        = RECONST_DTIMETRICS.out.pulsation_std_dwi
        pulsation_in_b0         = RECONST_DTIMETRICS.out.pulsation_std_b0
        dti_residual_stats      = RECONST_DTIMETRICS.out.residual_residuals_stats

        versions                = ch_versions                 // channel: [ path(versions.yml) ]
}
