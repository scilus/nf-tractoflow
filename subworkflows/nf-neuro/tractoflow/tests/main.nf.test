
nextflow_workflow {

    name "Test Subworkflow TRACTOFLOW"
    script "../main.nf"
    config "../nextflow.config"
    workflow "TRACTOFLOW"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/tractoflow"

    tag "load_test_data"
    tag "subworkflows/anatomical_segmentation"
    tag "subworkflows/preproc_dwi"
    tag "subworkflows/preproc_t1"
    tag "subworkflows/registration"

    tag "reconst/dtimetrics"
    tag "reconst/frf"
    tag "reconst/fodf"
    tag "reconst/meanfrf"
    tag "registration/antsapplytransforms"
    tag "tracking/localtracking"
    tag "tracking/pfttracking"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "antsbet.zip", "raw_DWIss1000-dir32.zip", "raw_T1w.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("tractoflow - no rev") {
        when {
            params {
                pft_number_of_seeds = 1
                pft_number_of_particles = 4
                lt_number_of_seeds = 1
            }
            workflow {
                """
                ch_split_test_data = LOAD_DATA.out.test_data_directory
                    .branch{
                        dwi: it.simpleName == "raw_DWIss1000-dir32"
                        t1w: it.simpleName == "raw_T1w"
                        template: it.simpleName == "antsbet"
                    }
                input[0] = ch_split_test_data.dwi.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]}
                input[1] = ch_split_test_data.t1w.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/T1w.nii.gz")
                    ]}
                input[2] = Channel.empty()
                input[3] = Channel.empty()
                input[4] = Channel.empty()
                input[5] = Channel.empty()
                input[6] = Channel.empty()
                input[7] = Channel.empty()
                input[8] = ch_split_test_data.template.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_template.nii.gz")
                    ]}
                input[9] = ch_split_test_data.template.map{
                    test_data_directory -> [
                        [ id:'test' ],
                        file("\${test_data_directory}/t1_brain_probability_map.nii.gz")
                    ]}
                input[10] = Channel.empty()
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                {
                    assert snapshot(
                                                                                                    // IN DIFFUSION SPACE
                        file(workflow.out.dwi.get(0).get(1)).name,                                  //  dwi
                        workflow.out.dwi.get(0).get(2),                                             //  bval
                        workflow.out.dwi.get(0).get(3),                                             //  bvec
                        file(workflow.out.t1.get(0).get(1)).name,                                   //  t1
                        file(workflow.out.wm_mask.get(0).get(1)).name,                              //  wm_mask
                        file(workflow.out.gm_mask.get(0).get(1)).name,                              //  gm_mask
                        file(workflow.out.csf_mask.get(0).get(1)).name,                             //  csf_mask
                        file(workflow.out.wm_map.get(0).get(1)).name,                               //  wm_map
                        file(workflow.out.gm_map.get(0).get(1)).name,                               //  gm_map
                        file(workflow.out.csf_map.get(0).get(1)).name,                              //  csf_map
                                                                                                    // REGISTRATION
                                                                                                    //  anatomical_to_diffusion
                        file(workflow.out.anatomical_to_diffusion.get(0).get(1)).name,              //   warp (or affine)
                        workflow.out.anatomical_to_diffusion.get(0).get(2)
                            ? file(workflow.out.anatomical_to_diffusion.get(0).get(2)).name : "",   //   affine
                                                                                                    //  diffusion_to_anatomical
                        file(workflow.out.diffusion_to_anatomical.get(0).get(1)).name,              //   affine (or warp)
                        workflow.out.diffusion_to_anatomical.get(0).get(2)
                            ? file(workflow.out.diffusion_to_anatomical.get(0).get(2)).name : "",   //   warp
                                                                                                    // IN ANATOMICAL SPACE
                        file(workflow.out.t1_native.get(0).get(1)).name,                            //  t1_native
                                                                                                    // DTI
                        file(workflow.out.dti_tensor.get(0).get(1)).name,                           //  dti_tensor
                        file(workflow.out.dti_md.get(0).get(1)).name,                               //  dti_md
                        file(workflow.out.dti_rd.get(0).get(1)).name,                               //  dti_rd
                        file(workflow.out.dti_ad.get(0).get(1)).name,                               //  dti_ad
                        file(workflow.out.dti_fa.get(0).get(1)).name,                               //  dti_fa
                        file(workflow.out.dti_rgb.get(0).get(1)).name,                              //  dti_rgb
                        file(workflow.out.dti_peaks.get(0).get(1)).name,                            //  dti_peaks
                        file(workflow.out.dti_evecs.get(0).get(1)).name,                            //  dti_evecs
                        file(workflow.out.dti_evals.get(0).get(1)).name,                            //  dti_evals
                        file(workflow.out.dti_residual.get(0).get(1)).name,                         //  dti_residual
                        file(workflow.out.dti_ga.get(0).get(1)).name,                               //  dti_ga
                        file(workflow.out.dti_mode.get(0).get(1)).name,                             //  dti_mode
                        file(workflow.out.dti_norm.get(0).get(1)).name,                             //  dti_norm
                                                                                                    // FODF
                        file(workflow.out.fiber_response.get(0).get(1)).name,                       //  fiber_response
                        file(workflow.out.fodf.get(0).get(1)).name,                                 //  fodf
                        file(workflow.out.fodf_peaks.get(0).get(1)).name,                           //  fodf_peaks
                        file(workflow.out.afd_max.get(0).get(1)).name,                              //  afd_max
                        file(workflow.out.afd_total.get(0).get(1)).name,                            //  afd_total
                        file(workflow.out.afd_sum.get(0).get(1)).name,                              //  afd_sum
                        file(workflow.out.nufo.get(0).get(1)).name,                                 //  nufo
                                                                                                    // TRACKING
                        file(workflow.out.pft_tractogram.get(0).get(1)).name,                       //  pft_tractogram
                        file(workflow.out.pft_config.get(0).get(1)).name,                           //  pft_config
                        file(workflow.out.pft_map_include.get(0).get(1)).name,                      //  pft_map_include
                        file(workflow.out.pft_map_exclude.get(0).get(1)).name,                      //  pft_map_exclude
                        file(workflow.out.pft_seeding_mask.get(0).get(1)).name,                     //  pft_seeding_mask
                        file(workflow.out.local_tractogram.get(0).get(1)).name,                     //  local_tractogram
                        file(workflow.out.local_config.get(0).get(1)).name,                         //  local_config
                        file(workflow.out.local_seeding_mask.get(0).get(1)).name,                   //  local_seeding_mask
                        file(workflow.out.local_tracking_mask.get(0).get(1)).name,                  //  local_tracking_mask
                                                                                                    // QC
                        file(workflow.out.nonphysical_voxels.get(0).get(1)).name,                   //  nonphysical_voxels
                        file(workflow.out.pulsation_in_dwi.get(0).get(1)).name,                     //  pulsation_in_dwi
                        file(workflow.out.pulsation_in_b0.get(0).get(1)).name,                      //  pulsation_in_b0
                        file(workflow.out.dti_residual_stats.get(0).get(1)).name,                   //  dti_residual_stats
                        workflow.out.versions                                                       // versions
                    ).match()
                }
            )
        }
    }
}
