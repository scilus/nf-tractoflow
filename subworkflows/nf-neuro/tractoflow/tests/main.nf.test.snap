{
    "tractoflow - no rev": {
        "content": [
            "test_dwi_resampled.nii.gz",
            "test__dwi_eddy_corrected.bval:md5,7995daabbd74fcf7c365a39a779f67e9",
            "test__dwi_eddy_corrected.bvec:md5,b3d6f6e2ec5c7cecad2207e5ea53fd79",
            "test__t1_warped.nii.gz",
            "test__mask_wm.nii.gz",
            "test__mask_gm.nii.gz",
            "test__mask_csf.nii.gz",
            "test__map_wm.nii.gz",
            "test__map_gm.nii.gz",
            "test__map_csf.nii.gz",
            "test__output1Warp.nii.gz",
            "test__output0GenericAffine.mat",
            "test__output0GenericAffine.mat",
            "test__output1InverseWarp.nii.gz",
            "test_t1_cropped.nii.gz",
            "test__tensor.nii.gz",
            "test__md.nii.gz",
            "test__rd.nii.gz",
            "test__ad.nii.gz",
            "test__fa.nii.gz",
            "test__rgb.nii.gz",
            "test__evecs_v1.nii.gz",
            "test__evecs.nii.gz",
            "test__evals.nii.gz",
            "test__residual.nii.gz",
            "test__ga.nii.gz",
            "test__mode.nii.gz",
            "test__norm.nii.gz",
            "test__frf.txt",
            "test__fodf.nii.gz",
            "test__peaks.nii.gz",
            "test__afd_max.nii.gz",
            "test__afd_total.nii.gz",
            "test__afd_sum.nii.gz",
            "test__nufo.nii.gz",
            "test__pft_tracking.trk",
            "test__pft_tracking_config.json",
            "test__map_include.nii.gz",
            "test__map_exclude.nii.gz",
            "test__pft_seeding_mask.nii.gz",
            "test__local_tracking.trk",
            "test__local_tracking_config.json",
            "test__local_seeding_mask.nii.gz",
            "test__local_tracking_mask.nii.gz",
            "test__nonphysical.nii.gz",
            "test__pulsation_std_dwi.nii.gz",
            "test__pulsation_std_b0.nii.gz",
            "test__residual_residuals_stats.png",
            [
                "versions.yml:md5,29eb4791c0e509ddbfa054f7d864e119",
                "versions.yml:md5,329641acd94a1eb6e65ef73b738b9165",
                "versions.yml:md5,4a7c801d46bade11523003b6bd06f8ed",
                "versions.yml:md5,647b63809e635c82552cf53178bf73be",
                "versions.yml:md5,9e058d5e2eaf4cf239570ca80d20207f",
                "versions.yml:md5,aaab00d1fa7d070cc7ee67fd768416b8",
                "versions.yml:md5,da10c87dbb2ad07aa8bbf89a0c172a60",
                "versions.yml:md5,e092829f98d53bda1802d1cda0cb017f",
                "versions.yml:md5,fc6a2d3923b6ecd60b833ce5f09f8877"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.10.4"
        },
        "timestamp": "2025-02-06T03:39:50.463608725"
    }
}